lib/wibox/container/radialprogressbar.lua

```diff
@@ -0,0 +1,264 @@
+---------------------------------------------------------------------------
+
+local setmetatable = setmetatable
+local base      = require("wibox.widget.base")
+local shape     = require("gears.shape"      )
+local util      = require( "awful.util"      )
+local color     = require( "gears.color"     )
+local beautiful = require("beautiful"        )
+
+local default_outline_width  = 2
+
+local radialprogressbar = { mt = {} }
+
+--- The progressbar border background color.
+-- @beautiful beautiful.radialprogressbar_border_color
+
+--- The progressbar foreground color.
+-- @beautiful beautiful.radialprogressbar_color
+
+--- The progressbar border width.
+-- @beautiful beautiful.radialprogressbar_border_width
+
+--- The padding between the outline and the progressbar.
+-- @beautiful beautiful.radialprogressbar_paddings
+-- @tparam[opt=0] table|number paddings A number or a table
+-- @tparam[opt=0] number paddings.top
+-- @tparam[opt=0] number paddings.bottom
+-- @tparam[opt=0] number paddings.left
+-- @tparam[opt=0] number paddings.right
+
+local function outline_workarea(self, width, height)
+    local border_width = self._private.border_width or
+        beautiful.radialprogressbar_border_width or default_outline_width
+
+    local x, y = 0, 0
+
+    -- Make sure the border fit in the clip area
+    local offset = border_width/2
+    x, y = x + offset, y+offset
+    width, height = width-2*offset, height-2*offset
+
+    return {x=x, y=y, width=width, height=height}, offset
+end
+
+-- The child widget area
+local function content_workarea(self, width, height)
+    local padding = self._private.paddings or {}
+    local wa = outline_workarea(self, width, height)
+
+    wa.x      = wa.x + (padding.left or 0)
+    wa.y      = wa.y + (padding.top  or 0)
+    wa.width  = wa.width  - (padding.left or 0) - (padding.right  or 0)
+    wa.height = wa.height - (padding.top  or 0) - (padding.bottom or 0)
+
+    return wa
+end
+
+-- Draw the radial outline and progress
+function radialprogressbar:after_draw_children(_, cr, width, height)
+    cr:restore()
+
+    local border_width = self._private.border_width or
+        beautiful.radialprogressbar_border_width or default_outline_width
+
+    local wa = outline_workarea(self, width, height)
+    cr:translate(wa.x, wa.y)
+
+    -- Draw the outline
+    shape.rounded_bar(cr, wa.width, wa.height)
+    cr:set_source(color(self:get_border_color() or "#0000ff"))
+    cr:set_line_width(border_width)
+    cr:stroke()
+
+    -- Draw the progress
+    cr:set_source(color(self:get_color() or "#ff00ff"))
+    shape.radial_progress(cr,  wa.width, wa.height, self._percent or 0)
+    cr:set_line_width(border_width)
+    cr:stroke()
+
+end
+
+-- Set the clip
+function radialprogressbar:before_draw_children(_, cr, width, height)
+    cr:save()
+    local wa = content_workarea(self, width, height)
+    cr:translate(wa.x, wa.y)
+    shape.rounded_bar(cr, wa.width, wa.height)
+    cr:clip()
+    cr:translate(-wa.x, -wa.y)
```

Comment:
I don't think this works correctly. The `wibox.hierarchy` now has a wrong idea on where the child widget is and so partial redraws will not redraw the correct part of the screen.

However, your `:layout()` already applies a translation (in fact, the opposite translation) to the child widget, so why is this needed at all?